service/monitoring_service.go

```diff
@@ -0,0 +1,156 @@
+package service
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/odpf/salt/log"
+
+	"github.com/odpf/optimus/models"
+	"github.com/odpf/optimus/store"
+)
+
+type monitoringService struct {
+	logger        log.Logger
+	JobRunService JobRunService
+
+	JobRunMetricsRepository store.JobRunMetricsRepository
+	TaskRunRepository       store.TaskRunRepository
+	SensorRunRepository     store.SensorRunRepository
+	HookRunRepository       store.HookRunRepository
+}
+
+type MonitoringService interface {
+	ProcessEvent(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec) error
+}
+
+func getSLADuration(jobSpec models.JobSpec) (int64, error) {
+	var slaMissDurationInSec int64
+	for _, notify := range jobSpec.Behavior.Notify {
+		if notify.On == models.SLAMissEvent {
+			if _, ok := notify.Config["duration"]; !ok {
+				continue
+			}
+
+			dur, err := time.ParseDuration(notify.Config["duration"])
+			if err != nil {
+				return 0, fmt.Errorf("failed to parse sla_miss duration %s: %w", notify.Config["duration"], err)
+			}
+			slaMissDurationInSec = int64(dur.Seconds())
+		}
+	}
+	return slaMissDurationInSec, nil
+}
+
+func (m monitoringService) registerNewJobRun(ctx context.Context, event models.JobEvent, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec) error {
+	slaMissDurationInSec, err := getSLADuration(jobSpec)
+	if err != nil {
+		return err
+	}
+	err = m.JobRunMetricsRepository.Save(ctx, event, namespaceSpec, jobSpec, slaMissDurationInSec)
+	if err != nil {
+		m.logger.Info(err.Error())
+		return err
+	}
+	return nil
+}
+
+func (m monitoringService) updateJobRun(ctx context.Context, event models.JobEvent, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec) error {
+	err := m.JobRunMetricsRepository.Update(ctx, event, namespaceSpec, jobSpec)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m monitoringService) getActiveJobRun(ctx context.Context, event models.JobEvent, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec) (models.JobRunSpec, error) {
+	eventPayload := event.Value
+	jobRunSpec, err := m.JobRunMetricsRepository.GetActiveJobRun(ctx, eventPayload["scheduled_at"].GetStringValue(), namespaceSpec, jobSpec)
```

Comment:
this would fail gracefully with time parse error for scheduled_at, as `eventPayload["scheduled_at"].GetStringValue()` shall give empty string when missing